Ubuntu Focal to Jammy migration support

seunghun1ee opened the pull request:
Add playbook and shell scripts to support automated Ubuntu migration.

Alex-Welsh committed:
Add playbook to upgrade ubuntu focal hosts

seunghun1ee committed:
add do-release-upgrade is installation task

Alex-Welsh committed:
Fix ubuntu upgrade playbook

Alex-Welsh committed:
Fix networking outage during ubuntu upgrade

seunghun1ee committed:
Add updating subset of ansible facts after do-release-upgrade

seunghun1ee committed:
Add shell script to automate ubuntu upgrade

seunghun1ee committed:
Add release note

seunghun1ee committed:
fix trailing space

technowhizz commented on etc/kayobe/ansible/ubuntu-upgrade.yml:
```diff
@@ -0,0 +1,112 @@
+        name: ubuntu-release-upgrader-core
+        state: latest
+      become: true
+
+    - name: Print selection of Deities to pray to
+      debug:
+        msg: |
+          The following Deities are associated with luck or good fortune:
+            Hamingja - Norse guardian spirit
+            Fortuna - Roman goddess of fortune and luck
+            Tyche - Greek goddess of success, fortune, Luck, and prosperity
+            Nang Kwak - Thai folklore goddess deemed to bring good fortune and prosperity
```
```suggestion
```

This repository is synced to and forked by multiple clients. We need to maintain our professionalism when doing so. This shouldn't be in here.

Alex-Welsh replied:
I'd say it's harmless fun, but if there's a consensus that it should be removed then we can get rid of it

seunghun1ee replied:
Do we remove this?

Alex-Welsh replied:
Maybe @markgoddard or @oneswig can weigh in

mnasiadka replied:
Does that mean the procedure is unreliable and you really need to pray? ;-)

Alex-Welsh replied:
It's pretty reliable, but if it fails then it's just a big pain

mnasiadka replied:
Big pain == reprovision the host? Maybe it's easier just to reprovision the host with Jammy image?

Alex-Welsh replied:
In-place has been better than reprovision. It's relatively fast and consistent. Maybe worth adding reprovision as a backup option to the guide though

technowhizz commented on etc/kayobe/ansible/ubuntu-upgrade.yml:
```diff
@@ -0,0 +1,112 @@
+        fail_msg: >-
+          This playbook is only designed for Ubuntu Focal 20.04 hosts. Ensure
+          that you are limiting it to only run on Focal hosts and
+          os_distribution in set to ubuntu.
```
```suggestion
          os_distribution is set to ubuntu.
```

seunghun1ee replied:
Oops nice catch

technowhizz commented on etc/kayobe/ansible/ubuntu-upgrade.yml:
```diff
@@ -0,0 +1,112 @@
+      assert:
+        that:
+          - ansible_facts.distribution == 'Ubuntu'
+          - ansible_facts.distribution_major_version == '20'
+          - ansible_facts.distribution_release == 'focal'
```
NIT: Not sure both of these are needed

seunghun1ee replied:
We wanted to be extra sure about this assert.

technowhizz replied:
Ubuntu 20 is always going to be Focal 😄 

But either way looks fine. Lets get this moving 👍

technowhizz commented on etc/kayobe/ansible/ubuntu-upgrade.yml:
```diff
@@ -0,0 +1,112 @@
+    - name: Assert that hosts are now using Ubuntu 22
+      assert:
+        that:
+          - ansible_facts.distribution_major_version == '22'
+          - ansible_facts.distribution_release == 'jammy'
```
NIT: Not sure both of these are needed

seunghun1ee replied:
Same reason with Focal assert.

technowhizz commented:
Overall, Good job guys @seunghun1ee @Alex-Welsh

Alex-Welsh commented on etc/kayobe/ansible/scripts/overcloud-ubuntu-upgrade.sh:
```diff
@@ -0,0 +1,34 @@
+
+kayobe playbook run $KAYOBE_CONFIG_PATH/ansible/ubuntu-upgrade.yml --limit $1
+
+kayobe overcloud host configure --limit $1
```
```suggestion
kayobe overcloud host configure --limit $1 --kolla-limit $1 -e os_release=jammy
```

technowhizz replied:
If you are on antelope when running that command then you will hit an error because `--kolla-limit` is no longer used with `host configure`. Although I'm guessing that this will probably not be the case and most likely will be on yoga?

Alex-Welsh replied:
There's no support for Focal on Zed+ so it must be on yoga at that point

seunghun1ee replied:
Added those changes

seunghun1ee committed:
Fix typo and praying

seunghun1ee committed:
Add kolla limit and override os_release to jammy as safety

technowhizz reviewed (dismissed):
LGTM 👍 

Thank you for the changes @seunghun1ee

markgoddard commented on etc/kayobe/ansible/scripts/seed-ubuntu-upgrade.sh:
```diff
@@ -0,0 +1,34 @@
+
+set -e
+
+if [[ ! $1 ]]; then
+    echo "Usage: seed-ubuntu-upgrade.sh <seed-hostname>"
+    exit 2
+fi
```
There should only be a single seed host

seunghun1ee replied:
But do they always have same host name?

markgoddard replied:
It doesn't matter - plays target the `seed` group

Alex-Welsh replied:
We don't need the hostname, because we never need to limit it

seunghun1ee replied:
I see

Alex-Welsh replied:
Done

markgoddard commented on etc/kayobe/ansible/scripts/overcloud-ubuntu-upgrade.sh:
We have a `tools/` directory that could be more appropriate for this

seunghun1ee replied:
Sounds good

Alex-Welsh replied:
Done

markgoddard commented on etc/kayobe/ansible/scripts/seed-ubuntu-upgrade.sh:
```diff
@@ -0,0 +1,34 @@
+
+kayobe playbook run $KAYOBE_CONFIG_PATH/ansible/ubuntu-upgrade.yml --limit $1
+
+kayobe seed host configure --limit $1
```
You don't have `-e os_release=jammy` here.

Alex-Welsh replied:
Added a precheck to the playbook to check it's set before running

markgoddard commented on etc/kayobe/ansible/ubuntu-upgrade.yml:
```diff
@@ -0,0 +1,103 @@
+---
+# To prevent Ansible role dependency errors, this playbook requires that environment variable
+# ANSIBLE_ROLES_PATH is defined and includes '$KAYOBE_PATH/ansible/roles' at seed-hypervisor.
```
```suggestion
# ANSIBLE_ROLES_PATH is defined and includes '$KAYOBE_PATH/ansible/roles' on the Ansible control host.
```

Alex-Welsh replied:
Done

markgoddard commented on etc/kayobe/ansible/ubuntu-upgrade.yml:
```diff
@@ -0,0 +1,103 @@
+# To prevent Ansible role dependency errors, this playbook requires that environment variable
+# ANSIBLE_ROLES_PATH is defined and includes '$KAYOBE_PATH/ansible/roles' at seed-hypervisor.
+- name: Migrate hosts from Ubuntu Focal 20.04 to Jammy 22.04
+  hosts: overcloud:infra-vms:seed
```
```suggestion
  hosts: overcloud:infra-vms:seed:seed-hypervisor
```

Alex-Welsh replied:
Done

markgoddard commented on etc/kayobe/ansible/scripts/seed-ubuntu-upgrade.sh:
Might want more scripts for infra VMs and the seed hypervisor?

Alex-Welsh replied:
Done

markgoddard commented on etc/kayobe/ansible/ubuntu-upgrade.yml:
```diff
@@ -0,0 +1,103 @@
+          deb {{ stackhpc_repo_ubuntu_jammy_url }} jammy main restricted universe multiverse
+          deb {{ stackhpc_repo_ubuntu_jammy_url }} jammy-updates main restricted universe multiverse
+          deb {{ stackhpc_repo_ubuntu_jammy_url }} jammy-backports main restricted universe multiverse
+          deb {{ stackhpc_repo_ubuntu_jammy_security_url }} jammy-security main restricted universe multiverse
```
Is the Jammy Docker repo not required?

Alex-Welsh replied:
Jammy docker gets sorted out by the host configure. This set is just what is required to get to Jammy.

markgoddard commented on etc/kayobe/ansible/ubuntu-upgrade.yml:
```diff
@@ -0,0 +1,103 @@
+
+    - name: Ensure do-release-upgrade is installed
+      package:
+        name: ubuntu-release-upgrader-core
```
Does this install the jammy version of this package? Is that a problem?

Alex-Welsh replied:
It shouldn't since the apt cache isn't updated, but I agree it looks kinda freaky so I've moved it

markgoddard commented on etc/kayobe/ansible/ubuntu-upgrade.yml:
```diff
@@ -0,0 +1,103 @@
+    - name: Ensure Jammy repo definitions do not exist in sources.list
+      blockinfile:
+        path: /etc/apt/sources.list
+        state: absent
```
FWIW, I think you can omit the block when setting state=absent, and avoid keeping it in sync. I don't know what it would do if it did not match the existing block. You can pass the `marker` argument to uniquely identify the block.

Alex-Welsh replied:
Done

markgoddard commented on etc/kayobe/ansible/ubuntu-upgrade.yml:
```diff
@@ -0,0 +1,103 @@
+
+    - name: Update Python and current user facts before re-creating Kayobe venv
+      ansible.builtin.setup:
+        gather_subset:
+          - '!all'
+          - python
+          - python_version
+          - user_uid
+          - user_gid
```
I'd avoid specifying a subset, in case anyone is using fact caching - it may not collect some facts that are later necessary. We have a variable for the subset for those who wish to tune it. There's also a filter that can be applied.
```suggestion
        filter: "{{ kayobe_ansible_setup_filter }}"
        gather_subset: "{{ kayobe_ansible_setup_gather_subset }}"
```

Alex-Welsh replied:
Done

markgoddard commented on etc/kayobe/ansible/ubuntu-upgrade.yml:
```diff
@@ -0,0 +1,103 @@
+- name: Run kayobe-target-venv playbook to ensure kayobe venv exists on remote host
+  import_playbook: "{{ lookup('ansible.builtin.env', 'VIRTUAL_ENV') }}/share/kayobe/ansible/kayobe-target-venv.yml"
+
+- name: Run network playbook to ensure networking config is not reliant on ifupdown
```
Ubuntu should never use ifupdown. I wonder why this is necessary.

seunghun1ee replied:
Hosts lost gateway definition after rebooting it to complete the upgrade. So, running the network playbook was necessary to keep the ssh connection with it. I think the name should be updated as it's not ifupdown was issue only when using ubuntu upstream as a source of do-release-upgrade.

Alex-Welsh replied:
I've updated the name of the play

markgoddard commented on etc/kayobe/ansible/ubuntu-upgrade.yml:
```diff
@@ -0,0 +1,103 @@
+  import_playbook: "{{ lookup('ansible.builtin.env', 'VIRTUAL_ENV') }}/share/kayobe/ansible/network.yml"
+
+- name: Reboot and confirm the host is upgraded to Jammy 22.04
+  hosts: overcloud:infra-vms:seed
```
```suggestion
  hosts: overcloud:infra-vms:seed:seed-hypervisor
```

Alex-Welsh replied:
Done

markgoddard commented on etc/kayobe/ansible/ubuntu-upgrade.yml:
```diff
@@ -0,0 +1,103 @@
+
+    - name: Update distribution facts
+      ansible.builtin.setup:
+        gather_subset:
```
As above

Alex-Welsh replied:
Done

markgoddard commented on etc/kayobe/ansible/ubuntu-upgrade.yml:
```diff
@@ -0,0 +1,103 @@
+- name: Reboot and confirm the host is upgraded to Jammy 22.04
+  hosts: overcloud:infra-vms:seed
+  vars:
+    ansible_python_interpreter: /usr/bin/python3
```
Shouldn't the Kayobe venv now be in place?

Alex-Welsh replied:
It should, but I think we hit some issues with it in testing. I can't remember precisely how it played out though

markgoddard reviewed (changes requested):
Would be nice to see some docs for this, but they can come later.

markgoddard reviewed (commented):
I think it would be sensible to set `os_release` to `jammy` at the beginning of the procedure, as we do for RL9. It's generally used as the 'target' release. This would avoid needing to pass in additional variables to pulp sync, host configure, etc.

Alex-Welsh commented:
> Would be nice to see some docs for this, but they can come later.

We have a wiki page for now. Docs are a bit awkward to update quickly. I'd like to have something more official in skc docs once it's stable though
https://wiki.stackhpc.com/doc/migrating-to-ubuntu-jammy-qtQNt9yw1m

Alex-Welsh committed:
Update ubuntu upgrade scripts after review

Alex-Welsh committed:
Rename ubuntu upgrade scripts

technowhizz commented:
@Alex-Welsh When approved and ready to merge, I think it would be a good idea to squash merge this PR so that it can easily be cherry picked

Alex-Welsh commented:
Error with latest compute node:
```
TASK [Do release upgrade] ***********************************************************************************************************************************************************************************************************│···································································································································································································
Monday 05 February 2024  11:46:20 +0000 (0:00:00.531)       0:00:58.421 *******                                                                                                                                                      │···································································································································································································
fatal: [ts-hcomp01]: FAILED! => changed=true                                                                                                                                                                                         │···································································································································································································
  cmd:                                                                                                                                                                                                                               │···································································································································································································
  - do-release-upgrade                                                                                                                                                                                                               │···································································································································································································
  - -f                                                                                                                                                                                                                               │···································································································································································································
  - DistUpgradeViewNonInteractive                                                                                                                                                                                                    │···································································································································································································
  delta: '0:00:02.624216'                                                                                                                                                                                                            │···································································································································································································
  end: '2024-02-05 11:46:23.248756'                                                                                                                                                                                                  │···································································································································································································
  msg: non-zero return code                                                                                                                                                                                                          │···································································································································································································
  rc: 1                                                                                                                                                                                                                              │···································································································································································································
  start: '2024-02-05 11:46:20.624540'                                                                                                                                                                                                │···································································································································································································
  stderr: ''                                                                                                                                                                                                                         │···································································································································································································
  stderr_lines: <omitted>                                                                                                                                                                                                            │···································································································································································································
  stdout: |-                                                                                                                                                                                                                         │···································································································································································································
    Checking for a new Ubuntu release                                                                                                                                                                                                │···································································································································································································
    You have not rebooted after updating a package which requires a reboot. Please reboot before upgrading.                                                                                                                          │···································································································································································································
  stdout_lines: <omitted> 
```
Not sure if it's the node or the change to the ordering

seunghun1ee commented:
> Error with updated ordering:
> 
> ```
> TASK [Do release upgrade] ***********************************************************************************************************************************************************************************************************│···································································································································································································
> Monday 05 February 2024  11:46:20 +0000 (0:00:00.531)       0:00:58.421 *******                                                                                                                                                      │···································································································································································································
> fatal: [ts-hcomp01]: FAILED! => changed=true                                                                                                                                                                                         │···································································································································································································
>   cmd:                                                                                                                                                                                                                               │···································································································································································································
>   - do-release-upgrade                                                                                                                                                                                                               │···································································································································································································
>   - -f                                                                                                                                                                                                                               │···································································································································································································
>   - DistUpgradeViewNonInteractive                                                                                                                                                                                                    │···································································································································································································
>   delta: '0:00:02.624216'                                                                                                                                                                                                            │···································································································································································································
>   end: '2024-02-05 11:46:23.248756'                                                                                                                                                                                                  │···································································································································································································
>   msg: non-zero return code                                                                                                                                                                                                          │···································································································································································································
>   rc: 1                                                                                                                                                                                                                              │···································································································································································································
>   start: '2024-02-05 11:46:20.624540'                                                                                                                                                                                                │···································································································································································································
>   stderr: ''                                                                                                                                                                                                                         │···································································································································································································
>   stderr_lines: <omitted>                                                                                                                                                                                                            │···································································································································································································
>   stdout: |-                                                                                                                                                                                                                         │···································································································································································································
>     Checking for a new Ubuntu release                                                                                                                                                                                                │···································································································································································································
>     You have not rebooted after updating a package which requires a reboot. Please reboot before upgrading.                                                                                                                          │···································································································································································································
>   stdout_lines: <omitted> 
> ```

This probably means that the host got newer version of kernel in current release.
As far as I know, rebooting with newly installed kernel will clear this issue.

Alex-Welsh committed:
Improved ubuntu upgrade script

Reboot prepatory apt update when required and always set -e
os_release=jammy for the upgrade playbook

seunghun1ee committed:
Trim trailing whitespaces

Alex-Welsh committed:
Run OVN chassis playbook after Jammy upgrade

seunghun1ee commented on etc/kayobe/ansible/ubuntu-upgrade.yml:
```diff
@@ -100,3 +100,7 @@
         that:
           - ansible_facts.distribution_major_version == '22'
           - ansible_facts.distribution_release == 'jammy'
+
+- name: Run the OVN chassis priority fix playbook
+  import_playbook: "{{ lookup('ansible.builtin.env', 'KAYOBE_CONFIG_PATH') }}/ansible/ovn-fix-chassis-priorities.yml"
+  when: kolla_enable_ovn
```
Do we need to run this playbook on all environments?

Alex-Welsh replied:
Only environments using OVN I think

Alex-Welsh committed:
Improve ubuntu upgrade playbook reliability

markgoddard commented on tools/ubuntu-upgrade-infra-vm.sh:
```diff
@@ -0,0 +1,34 @@
+
+kayobe playbook run $KAYOBE_CONFIG_PATH/ansible/ubuntu-upgrade.yml -e os_release=jammy --limit $1
+
+kayobe infra vm host configure --limit $1 --kolla-limit $1 -e os_release=jammy
```
This one doesn't use kolla (so this would fail)
```suggestion
kayobe infra vm host configure --limit $1 -e os_release=jammy
```

seunghun1ee committed:
Remove kolla limit for infra vm host configure

Co-authored-by: Mark Goddard <mark@stackhpc.com>
